packages/beacon-node/src/chain/seenCache/seenBlockInput.ts

```diff
@@ -0,0 +1,253 @@
+ *   https://github.com/ChainSafe/lodestar/blob/unstable/packages/beacon-node/src/chain/blocks/verifyBlocksDataAvailability.ts#L111
+ */
+
+export class SeenBlockInputCache {
+  private readonly config: ChainForkConfig;
+  private readonly clock: IClock;
+  private readonly chainEvents: ChainEventEmitter;
+  private readonly signal: AbortSignal;
+  private readonly metrics: Metrics | null;
+  private readonly logger?: Logger;
+  private blockInputs = new Map<RootHex, IBlockInput>();
+
+  constructor({config, clock, chainEvents, signal, metrics, logger}: SeenBlockInputCacheModules) {
+    this.config = config;
+    this.clock = clock;
+    this.chainEvents = chainEvents;
+    this.signal = signal;
+    this.metrics = metrics;
+    this.logger = logger;
+
+    if (metrics) {
+      metrics.seenCache.blockInput.blockInputCount.addCollect(() =>
+        metrics.seenCache.blockInput.blockInputCount.set(this.blockInputs.size)
+      );
+    }
+
+    this.chainEvents.on(ChainEvent.forkChoiceFinalized, this.onFinalized);
+    this.signal.addEventListener("abort", () => {
+      this.chainEvents.off(ChainEvent.forkChoiceFinalized, this.onFinalized);
+    });
+  }
+
+  has(rootHex: RootHex): boolean {
+    return this.blockInputs.has(rootHex);
+  }
+
+  get(rootHex: RootHex): IBlockInput | undefined {
+    return this.blockInputs.get(rootHex);
+  }
+
+  remove(rootHex: RootHex): void {
+    this.blockInputs.delete(rootHex);
+  }
+
+  prune(rootHex: RootHex): void {
+    let blockInput = this.blockInputs.get(rootHex);
+    let parentRootHex = blockInput?.parentRootHex;
+    while (blockInput) {
+      this.blockInputs.delete(blockInput.blockRootHex);
+      blockInput = this.blockInputs.get(parentRootHex ?? "");
+      parentRootHex = blockInput?.parentRootHex;
+    }
+  }
+
+  onFinalized = (checkpoint: CheckpointWithHex) => {
+    const cutoffSlot = computeStartSlotAtEpoch(checkpoint.epoch);
+    for (const [rootHex, blockInput] of this.blockInputs) {
+      if (blockInput.slot < cutoffSlot) {
+        this.blockInputs.delete(rootHex);
+      }
+    }
+  };
+
+  getByBlock({block, source, seenTimestampSec, peerIdStr}: SourceMeta & {block: SignedBeaconBlock}): IBlockInput {
+    const blockRoot = this.config.getForkTypes(block.message.slot).BeaconBlock.hashTreeRoot(block.message);
+    const blockRootHex = toRootHex(blockRoot);
+
+    // TODO(peerDAS): Why is it necessary to static cast this here. All conditional paths result in a valid value so should be defined correctly below
+    let blockInput = this.blockInputs.get(blockRootHex) as IBlockInput;
+    if (!blockInput) {
+      const {forkName, daOutOfRange} = this.buildCommonProps(block.message.slot);
+      if (!isForkPostDeneb(forkName)) {
+        blockInput = BlockInputPreData.createFromBlock({
+          block,
+          blockRootHex,
+          daOutOfRange,
+          forkName,
+          source: {
+            source,
+            seenTimestampSec,
```

Comment:
looks like `BlockInputBlobs` only tracks `timeCompleteSec`, would be great to also add `timeCreatedSec`, track it and add both to `getLogMeta()`, that would help us debug easier later especially for fulu